### ISSUE TITLE
StackOverflowError in deserialization

### ISSUE THREAD

**Issue by khogeland** [FasterXML/jackson-databind#1502]
Encountered while attempting to reproduce https://github.com/FasterXML/jackson-databind/issues/1501, possibly related.

Minimal repro:
```java
public class Something {
    @JsonProperty
    public InnerSomething a;

    @JsonCreator
    public Something(InnerSomething a) {}

    class InnerSomething {
        @JsonCreator
        public InnerSomething() {}
    }

}
```
```java
        ObjectMapper mapper = new ObjectMapper();
        String ser = mapper.writeValueAsString(new Something(null));
        mapper.readValue(ser, Something.class);
```

Produces:
```
java.lang.StackOverflowError
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:142)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:135)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1119)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:298)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:168)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:135)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1119)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:298)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:168)
```

**Comment by cowtowncoder** [comment]
Ok, nope. Fix for #1501 didn't handle this case yet.

**Comment by cowtowncoder** [comment]
Interesting. What happens at low level is a tight loop of 2 deserializers, one for inner, one for outer class. This sort of makes sense when realizing that "hidden" this point makes constructor for inner class take instance of outer class, and with `@JsonCreator` it is then assumed that it's delegating... and this leads to infinite ping pong via deserializers (since such recursive relationship are properly resolved), because no input is actually consumed.

I will prevent this (in 2.7 at least) by simply removing handling of constructors for static inner classes, except for special case of "no args" (except for implicit parent this) which is not even handled via creators but as special case of a child property.
Long story short, this will prevent SO, and produce some other exception that is more likely to point to actual problem (of trying to deserialize non-static inner class).
And for 2.8 I will see if I can improve actual exception message further.
